cvmfs/swissknife_filestats.h

```diff
@@ -0,0 +1,86 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#ifndef CVMFS_SWISSKNIFE_FILESTATS_H_
+#define CVMFS_SWISSKNIFE_FILESTATS_H_
+
+#include "swissknife.h"
+
+#include <pthread.h>
+
+#include <string>
+
+#include "catalog_traversal.h"
+#include "sql.h"
+
+using namespace std;  // NOLINT
+
+namespace swissknife {
+
+// Database class for storing repo statistics, the class design
+// follows StatisticsDatabase's design
+class FileStatsDatabase : public sqlite::Database<FileStatsDatabase> {
+ public:
+  bool CreateEmptyDatabase();
+
+  // Following Store* functions return the primary key value of the
+  // inserted entry
+  int StoreCatalog(int num_entries, int file_size);
+  int StoreFile(int catalog_id, int object_id);
+  int StoreObject(const void *hash, int hash_size, int64_t size);
+  int StoreChunkedFile(int catalog_id);
+  int StoreChunk(const void *hash, int hash_size, int64_t size, int file_id);
+  int StoreSymlink(int size);
+  void InitStatements();
+  void DestroyStatements();
+
+  static float kLatestSchema;
+  static unsigned kLatestSchemaRevision;
+
+ protected:
+  sqlite::Sql *query_insert_catalog;
+  sqlite::Sql *query_insert_object;
+  sqlite::Sql *query_insert_file;
+  sqlite::Sql *query_insert_file_object;
+  sqlite::Sql *query_insert_symlink;
+  sqlite::Sql *query_lookup_object;
+  friend class sqlite::Database<FileStatsDatabase>;
+  FileStatsDatabase(const std::string &filename,
+                    const OpenMode     open_mode)
+    : sqlite::Database<FileStatsDatabase>(filename, open_mode) { }
+};
+
+
+class CommandFileStats : public Command {
+ public:
+  ~CommandFileStats() {}
+  virtual std::string GetName() const { return "filestats"; }
+  virtual std::string GetDescription() const {
+    return "CernVM File System repository statistics exporter.";
+  }
+  virtual ParameterList GetParams() const;
+  int Main(const ArgumentList &args);
+
+ protected:
+  string db_path_;
+  string tmp_db_path_;
+  FileStatsDatabase* db_;
+
+  pthread_t thread_processing_;
+  volatile int num_downloaded_;
+  volatile bool running_;
```

Comment:
For multi-threading safety, it would be better to use atomic integers (see atomic.h)